### ISSUE TITLE
Iterating over project tracks retrieves attributes only for the current project

### ISSUE THREAD

**Issue by DrewTNBD** [RomeoDespres/reapy#80]
I'm not sure if this is a bug or if I am using the API incorrectly. But when I iterate over projects and their tracks, I lose access to track attributes such as name, items, etc.

Here is some lua code:
```
projects = {}

for i = 0, 10 do
  retval, projfn = reaper.EnumProjects(i)
  if retval ~= nil then
  projects[i] = {}
  projects[i][i] = retval
  end
end

projectCount = tableLength(projects) -- just gets the number of elements in the project table

for x = 0, projectCount -1 do
  projecttrackCount = reaper.CountTracks(projects[x][x])
  projects[x].trackCount = projecttrackCount
end

index = 0
for x = 0, projectCount -1 do
  trackCount = projects[x].trackCount
  
  for y = 0, trackCount -1 do
    track = reaper.GetTrack(projects[x][x], y)
    retval, trackname = reaper.GetSetMediaTrackInfo_String(track, "P_NAME", "string", false )
    Msg(trackname)
  end
  
end
```

What this does is gets the number of projects, gets the number of tracks in each project, and stores to a table. Then I iterate over each project, then each track and retrieve the name of the track and Msg out to the Reaper console. It's long-winded, but this shows that in Lua I don't have to be aware of the current project setting.

Here is some (simpler!!) reapy python code:

```
import reapy as rp


with rp.inside_reaper():
    for proj in rp.get_projects():
        print([s.name for s in proj.tracks])
```

I am attempting to do exactly the same here by iterating over every project and using the OOB abilities of reapy to retrieve the track names. However I only ever get the track names for whichever project is currently active in reaper.

Is this a bug, or am I misunderstanding the API?

**Comment by RomeoDespres** [comment]
It's actually a bug from ReaScript: accessing track names only works for current project.

With a bit of overhead, a solution would be:

```python
with rp.inside_reaper():
    for proj in rp.get_projects():
        with proj.make_current_project():
            print([s.name for s in proj.tracks])

**Comment by DrewTNBD** [comment]
I thought that too. But then I tried:
```
with rp.inside_reaper():
    for proj in rp.get_projects():
        print([s.items for s in proj.tracks])
```

And what I get is for the first project (the _current_ one) is an array of tracks, each with media items on them. But for the second project I get an array of empty arrays. It doesn't get the items in the project even though they are there.

```
[[Item("(MediaItem*)0x00007F986D5EF420"), Item("(MediaItem*)0x00007F986D5D4970"), Item("(MediaItem*)0x00007F986D5DED70"), Item("(MediaItem*)0x00007F986D5DC8D0"), Item("(MediaItem*)0x00007F9872D2B800"), Item("(MediaItem*)0x00007F9872D2C300"), Item("(MediaItem*)0x00007F9872C5A440"), Item("(MediaItem*)0x00007F9872C5D800"), Item("(MediaItem*)0x00007F9872D3A0D0"), Item("(MediaItem*)0x00007F9872D2CBE0"), Item("(MediaItem*)0x00007F9872D2D0F0"), Item("(MediaItem*)0x00007F9872D2D600"), Item("(MediaItem*)0x00007F9872D2DB10"), Item("(MediaItem*)0x00007F9872D2E020"), Item("(MediaItem*)0x00007F9875BE6A40"), Item("(MediaItem*)0x00007F986D771000"), Item("(MediaItem*)0x00007F9875820380"), Item("(MediaItem*)0x00007F9870830230"), Item("(MediaItem*)0x00007F9875C1D890"), Item("(MediaItem*)0x00007F9872D2E530"), Item("(MediaItem*)0x00007F9872D3C860"), Item("(MediaItem*)0x00007F9875CADB40"), Item("(MediaItem*)0x00007F9875CFE240"), Item("(MediaItem*)0x00007F9875C8A280"), Item("(MediaItem*)0x00007F9875C89CB0"), Item("(MediaItem*)0x00007F986D770740"), Item("(MediaItem*)0x00007F987583AE70"), Item("(MediaItem*)0x00007F9875B705E0"), Item("(MediaItem*)0x00007F9872C63490"), Item("(MediaItem*)0x00007F9872C637A0"), Item("(MediaItem*)0x00007F9872C64080"), Item("(MediaItem*)0x00007F9872D3D8E0")], [Item("(MediaItem*)0x00007F986D7FD0A0"), Item("(MediaItem*)0x00007F986D7FD5B0"), Item("(MediaItem*)0x00007F986D7FDAC0"), Item("(MediaItem*)0x00007F986D7FE510"), Item("(MediaItem*)0x00007F986D7FEA20"), Item("(MediaItem*)0x00007F986D7FEF30"), Item("(MediaItem*)0x00007F986D7FF440"), Item("(MediaItem*)0x00007F986D7FF950"), Item("(MediaItem*)0x00007F986D7FA380"), Item("(MediaItem*)0x00007F986D7FA690"), Item("(MediaItem*)0x00007F986D7FA9A0"), Item("(MediaItem*)0x00007F986D7FACB0"), Item("(MediaItem*)0x00007F986D7FAFC0"), Item("(MediaItem*)0x00007F986D7FB2D0"), Item("(MediaItem*)0x00007F986D7FB7E0"), Item("(MediaItem*)0x00007F986D7FBCF0"), Item("(MediaItem*)0x00007F986D7FC680"), Item("(MediaItem*)0x00007F986D7FCB90"), Item("(MediaItem*)0x00007F986D7C50A0"), Item("(MediaItem*)0x00007F986D7C53B0"), Item("(MediaItem*)0x00007F986D7F9740"), Item("(MediaItem*)0x00007F986D7F9A50"), Item("(MediaItem*)0x00007F986D7F9D60"), Item("(MediaItem*)0x00007F986D7FA070"), Item("(MediaItem*)0x00007F986D7D2C80"), Item("(MediaItem*)0x00007F986D7D3410"), Item("(MediaItem*)0x00007F986D7D3720"), Item("(MediaItem*)0x00007F986D7D3C30"), Item("(MediaItem*)0x00007F986D7D4140"), Item("(MediaItem*)0x00007F986D7D4650"), Item("(MediaItem*)0x00007F986D7D4B60"), Item("(MediaItem*)0x00007F986D7D5070")], [], []]
[[], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []]

```

If I change my _current_ project to the second one, then the second returns items fine but the first does not. So:

```
[[], [], [], []]
[[Item("(MediaItem*)0x00007F987082B640")], [Item("(MediaItem*)0x00007F9875B954D0")], [Item("(MediaItem*)0x00007F986D466110")], [Item("(MediaItem*)0x00007F9875B2BD00")], [Item("(MediaItem*)0x00007F9875B2C470")], [Item("(MediaItem*)0x00007F9875B2C980")], [Item("(MediaItem*)0x00007F9875B1A310")], [Item("(MediaItem*)0x00007F9872D26F40")], [Item("(MediaItem*)0x00007F9875BC8A70")], [Item("(MediaItem*)0x00007F9875B272D0")], [Item("(MediaItem*)0x00007F9875B27A40")], [Item("(MediaItem*)0x00007F986D5A50C0")], [Item("(MediaItem*)0x00007F9875B28410")], [Item("(MediaItem*)0x00007F9875B31480")], [Item("(MediaItem*)0x00007F9875B31790")], [Item("(MediaItem*)0x00007F9875B26B40")], [Item("(MediaItem*)0x00007F987082B950")], [Item("(MediaItem*)0x00007F9875B327E0")], [Item("(MediaItem*)0x00007F9875BFBE70")], [Item("(MediaItem*)0x00007F9875BFC380")], [Item("(MediaItem*)0x00007F9875BFC890")]]
```

**Comment by DrewTNBD** [comment]
Maybe it's because s.items for s in proj.tracks works on a track-name basis?

**Comment by Levitanus** [comment]
Actually, we can check whether it is an internal REAPER bug, or bug of the python C-bindings.
If you can make lua code, which consider project not as int index, but as project type and iterate over them — it will be interesting to look at.
Maybe, if it is not core REAPER behavior bug — I can fix it on days.

Sorry, can't test by myself, too busy with midi_events

**Comment by DrewTNBD** [comment]
@Levitanus Apologies, I'm not sure what you mean by 

>  consider project ... as project type

Could you please clarify? I'm happy to work on some more example lua to see if this is a bug internal to Reaper or a bug with the python C-bindings.

**Comment by DrewTNBD** [comment]
@Levitanus 
Okay, I think you meant this. If not, apologies.
```
projects = {}

for i = 0, 10 do
  retval, projfn = reaper.EnumProjects(i)
  if retval ~= nil then
  projects[i] = retval
  end
end

projectCount = tableLength(projects) -- just gets the number of elements in the project table

index = 0
for key, project in spairs(projects) do
  projectTrackCount = reaper.CountTracks(project)
  for y = 0, projectTrackCount -1 do
    track = reaper.GetTrack(project, y)
    retval, trackname = reaper.GetSetMediaTrackInfo_String(track, "P_NAME", "string", false )
    Msg(trackname)
  end
end
```

This gets project objects, puts them in a table, then for each one counts the tracks, then iterates around that number of tracks to get the track and retrieve the name. It works for all open projects.

If this isn't what you meant, please let me know what you mean by "project type" because so far as I can tell the only way to enumerate the open projects is by using indexes in the first loop to get the projects.

**Comment by Levitanus** [comment]
Yep. I mean this. So, this is definitely bug of Python bindings. I'll investigate on a week, but cannot promise quick results...
Use `make_current_project` for a while

**Comment by DrewTNBD** [comment]
Awesome, thanks for confirming.

**Comment by Levitanus** [comment]
OK, I've found the reason, but I need time to think, what to do with it:

When packing pointer to be passed into the C-function (for example `GetTrackName`), it is validated by API call `ValidatePtr`, which is project-independednt. So, when the pointer is not from the active project it is NULL. Otherwise, there is `ValidatePtr2`, which accepts project as argument, but rewriting API for be project-depended, and include everything is needed to that calls... Scary, at least)))

So, there is option not to check `ptr` for being valid. I can't imagine what it can lead to.

```Python
def packp(t, v):
    print(f'packp(t={t}, v={v})')
    m = re.match('^\((\w+\*|HWND)\)0x([0-9A-F]+)$', str(v))
    if (m != None):
        (_t, _v) = m.groups()
        if (_t == t or t == 'void*'):
            print(f'---- _t={_t}, _v={_v}')
            a = int(_v[:8], 16)
            b = int(_v[8:], 16)
            p = ct.c_uint64((a << 32) | b).value
            print(f'---- a={a}, b={b}, p={p}')
            if (ValidatePtr(p, t)):
                return p
            return p # works fine with this line
    return 0
```

**Comment by Levitanus** [comment]
I've included a simple fix in #77. It still seemed to me as dirty hack, but I've checked it within my tests (more than 30 cases, including managing of 5 projects within subclassing of `core` classes), and nothing has failed, so, lets consider this as «pretty safe» :)

**Comment by RomeoDespres** [comment]
Hi, just wanted to let you know that the fix from #77 has been merged into master and released on PyPI.
